### PR TITLE
`formatDateTime()`: Make placeholder `%f` more compatible with MySQL

### LINKED ISSUE #72879: formatDateTime %f inconsistent depending on scale/precision
**Company or project name**

Cribl

**Describe what's wrong**

When using `formatDateTime()`, the `%f` format doesn't conform to MySQL's definition.  It's supposed to be "Microseconds (000000..999999)" but the length seems to vary based on the scale/precision of the DateTime.

Easily reproduced:
https://fiddle.clickhouse.com/d17eb9b5-6cb0-4a7b-81c1-8163c799761c

**Does it reproduce on the most recent release?**

Yes

**How to reproduce**

Doesn't matter which server version/interface/settings.  See the Fiddle link above, or just run:

```sql
select
    formatDateTime(toDateTime64(now(), 2), '%f'),
    formatDateTime(toDateTime64(now(), 3), '%f'),
    formatDateTime(toDateTime64(now(), 4), '%f'),
    formatDateTime(now(), '%f')
```

**Expected behavior**

I'd expect 6 digits every time irrespective of the date's precision.  i.e. `000000` just as MySQL documented it.

### PR DESCRIPTION
**Pull request by ollidraese**
Resolves #72879

`%f` originally only produced as many digits (for the sub-section fraction) as specified by the scale of the date/time type. In contrast, MySQL unconditionally produces six digits and uses the scale only to define which N left-most digits are filled with values.

For example, in MySQL `SELECT DATE_FORMAT('2009-10-04 22:23:00.1234', '%f')` returns

```
+-----------------------------------------------+
| DATE_FORMAT('2009-10-04 22:23:00.1234', '%f') |
+-----------------------------------------------+
| 123400                                        |
+-----------------------------------------------+
```

As per this PR, `%f` now behaves the same.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The `%f` placeholder of function `formatDateTime` now unconditionally generates six (sub-second) digits. This makes the behavior compatible with MySQL `DATE_FORMAT` function. The previous behavior can be restored using setting `formatdatetime_f_prints_scale_number_of_digits = 1`.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 9374a47c7a58fb46fcfcf6d63bb6f8f4617f2a7c with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73324/9374a47c7a58fb46fcfcf6d63bb6f8f4617f2a7c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/7089511d1413e7098cc34f1a090cf23315581fea/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Functions/formatDateTime.cpp [hunk 1]
```diff
@@ -497,15 +497,20 @@ class FunctionFormatDateTimeImpl : public IFunction
...
+            // if more than 6 digits are passed in, we truncate to 6 == buffer size
+            while (scale > 6)
+            {
+                --scale;
+                fractional_second /= 10;
+            }
 
-            for (Int64 i = scale, value = fractional_second; i > 0; --i)
+            // mysql format option %f always produces six digits. scale only defines which of them get filled
+            for (UInt32 i = scale, value = fractional_second; i > 0 && value > 0; --i)
```
Comment (antaljanosbenjamin) [inline src/Functions/formatDateTime.cpp:508]: Should we fill all 6 digits then? If scale is less than 6 we might have some bits unitiailized, I think it makes sense to fill those places out also (with `0`).

**Review by antaljanosbenjamin — DISMISSED** [review]
I think you should introduce a settings to be able to keep the old behavior, as some people might depend on it. See [these](https://github.com/ClickHouse/ClickHouse/blob/d0f439d26e26b6caaef9beaf0c818624df3b33ea/src/Core/SettingsChangesHistory.cpp#L453-L454) settings that were introduced by [this PR](https://github.com/ClickHouse/ClickHouse/pull/47246).

File: src/Functions/formatDateTime.cpp (hunk 1 above)
Comment (ollidraese) [reply src/Functions/formatDateTime.cpp:508]: The buffer is initialized with 6 '0' characters (see line 1238 below). So, if we stop early here (i.e. value is zero), we keep the '0' in place. Note that this is part of the test-cases where I use scale < 6.

**Review by antaljanosbenjamin — COMMENTED** [review]
I think you still need to add a setting to control this behavior, see my comment above.

**Comment by ollidraese** [comment]
@antaljanosbenjamin, I can certainly add another option, but I would like to point out that this means that Clickhouse would have two options for controlling a single %f format option.

1. Do completely wrong (single zero) - old behavior
2. Somehow wrong (only generate digits for specified scale, not what the documentation is saying and MySQL is doing for years)
3. With this PR: Doing right (behave as %f is documented for a long time).

I personally find these choices a little excessive and only add burden on the admin side, but I'm happy to add it.

**Comment by antaljanosbenjamin** [comment]
@ollidraese I understood your concerns and I am thinking almost the same. Just to be clear: I think the new, correct behaviour should be the default.

We have a compatbility setting that ensures that even with upgrades users can keep the old behavior whenever it is possible. In this case the change is a breaking change (some people might depend on the current formatting), but it is not a "disasterous" bug, so it doesn't affect the correctness of the queries (as long as the new format is respected) or cause any permanent issues. The PR I linked is very similar to this one, so I think it is good to follow the convention.

**Comment by ollidraese** [comment]
@antaljanosbenjamin makes sense! I will add an option (probably with a little holiday delay). Thx!

**Comment by ollidraese** [comment]
@antaljanosbenjamin I added a new formatdatetime_f_prints_scale_only setting (defaulting to false) that restores the behavior before the PR. I assume this needs documentation, but I am unsure if I should update ChangeLog and/or doc (assuming there is a dedicated docs team). Also altered the tests to use both variants of the setting.

File: src/Core/Settings.cpp [hunk 2]
```diff
@@ -3148,6 +3148,9 @@ Allow execute multiIf function columnar
 )", 0) \
     DECLARE(Bool, formatdatetime_f_prints_single_zero, false, R"(
 Formatter '%f' in function 'formatDateTime' prints a single zero instead of six zeros if the formatted value has no fractional seconds.
+)", 0) \
```
Comment (rschu1ze) [inline src/Core/Settings.cpp:3151]: Let's also update the "Notes" below the table in the docs of function "formatDateTime": https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#formatdatetime

Comment (ollidraese) [reply src/Core/Settings.cpp:3151]: Thanks. Done.

**Comment by rschu1ze** [comment]
ClickHouse Integration tests (aarch64) [1/6]:
- test_memory_limit/test.py::test_multiple_queries: #72458

ClickHouse Stateless tests (debug, s3 storage):
- 02962_system_sync_replica_lightweight_from_modifier: #60480

**Pull request merged**